pkg/services/notary/notary.go

```diff
@@ -0,0 +1,364 @@
+										invScript = append(invScript, sig...)
+									}
+								}
+								r.main.Scripts[i].InvocationScript = invScript
+							}
+							break loop
+						}
+					}
+					// pubKey was not found for the signature i.e. signature is bad - we're OK with that, let the fallback TX to be added
+					break loop // only one multisignature is allowed
+				}
+			}
+		}
+	}
+	var isPresent bool
+	for _, f := range r.fallbacks {
+		if f.Hash().Equals(payload.FallbackTransaction.Hash()) {
+			isPresent = true
+			break
+		}
+	}
+	if !isPresent {
+		r.fallbacks = append(r.fallbacks, payload.FallbackTransaction)
+	}
+	if r.nSigsCollected == nSigs && r.isValid && r.minNotValidBefore > n.Config.Chain.BlockHeight() {
+		if err := n.finalize(r.main); err != nil {
+			n.Config.Log.Error("failed to finalize main transaction", zap.Error(err))
+		} else {
+			r.isSent = true
+		}
+	}
+	n.requests[payload.MainTransaction.Hash()] = r
+}
+
+// OnFallbackRemoval is a callback which is called after fallback transaction is removed
+// from the notary payload pool due to expiration, main tx appliance or any other reason.
+func (n *Notary) OnFallbackRemoval(pld *payload.P2PNotaryRequest) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	r, ok := n.requests[pld.MainTransaction.Hash()]
+	if !ok {
+		return
+	}
+	for i, fb := range r.fallbacks {
+		if fb.Hash().Equals(pld.FallbackTransaction.Hash()) {
+			r.fallbacks = append(r.fallbacks[:i], r.fallbacks[i+1:]...)
+			break
+		}
+	}
+	if len(r.fallbacks) == 0 {
+		delete(n.requests, r.main.Hash())
+	} else {
+		n.requests[pld.MainTransaction.Hash()] = r
+	}
+}
+
+// PostPersist is a callback which is called after new block is persisted.
+func (n *Notary) PostPersist(bc blockchainer.Blockchainer, pool *mempool.Pool, b *block.Block) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	for _, tx := range b.Transactions { // TODO: this cycle can be removed, because we have the same logic in OnFallbackRemoval
+		if _, ok := n.requests[tx.Hash()]; ok {
+			delete(n.requests, tx.Hash()) // main transaction is accepted, so request can be removed
+		}
+	}
+	for h, r := range n.requests {
+		if !r.isSent && r.isValid && r.nSigs == r.nSigsCollected && r.minNotValidBefore > bc.BlockHeight() {
+			if err := n.finalize(r.main); err != nil {
+				n.Config.Log.Error("failed to finalize main transaction", zap.Error(err))
+			} else {
+				r.isSent = true
+			}
+			continue
+		}
+		if r.minNotValidBefore <= bc.BlockHeight() { // then at least one of the fallbacks can already be sent.
+			newFallbacks := r.fallbacks[:0]
+			for _, fb := range r.fallbacks {
+				if nvb := fb.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height; nvb <= bc.BlockHeight() {
+					if err := n.finalize(fb); err != nil {
+						newFallbacks = append(newFallbacks, fb) // wait for the next block to resend them
```

Comment:
We might as well get to VUB waiting for successful attempt, so this should also be accounted for.

Reply:
Main and fallback have the same VUB. If VUB is reached, then fallback will be removed from the payload pool and, thanks to the pool callbacks, the request will be removed from the module. We perform all requests-related operations under the lock, so I think that it's not the module's responsibility to keep an eye on VUB.